File: python_solutions/1570.dot-product-of-two-sparse-vectors.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right
from functools import reduce, lru_cache
from typing import List
import itertools
import math
import heapq
import string
true = True
false = False
MIN, MAX, MOD = -0x3f3f3f3f, 0x3f3f3f3f, 1000000007


# https://leetcode.com/problems/dot-product-of-two-sparse-vectors/description/
# Medium
class SparseVector:
    def __init__(self, nums: List[int]):
        self.cc = []
        for i, n in enumerate(nums):
            if n > 0:
                self.cc.append((i, n))

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec: 'SparseVector') -> int:
        i = j = res = 0
        while i < len(self.cc) and j < len(vec.cc):
            if self.cc[i][0] == vec.cc[j][0]:
                res += self.cc[i][1] * vec.cc[j][1]
                i += 1
                j += 1
            elif self.cc[i][0] < vec.cc[j][0]:
                i += 1
            elif self.cc[i][0] > vec.cc[j][0]:
                j += 1
        return res


# Your SparseVector object will be instantiated and called as such:
# v1 = SparseVector(nums1)
# v2 = SparseVector(nums2)
# ans = v1.dotProduct(v2)

File: python_solutions/1570.dot-product-of-two-sparse-vectors.cpp
// Containers
#include <deque>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <iostream>
#include <numeric>



using namespace std; 
int dirs[5] = {-1, 0, 1, 0, -1};

// --------------------------------------------------------
class SparseVector {
public:
    std::vector<pair<int, int>> cc; 
    SparseVector(vector<int> &nums) {
        for(int i=0; i<nums.size(); i++){
          if (nums[i]>0) cc.push_back({i, nums[i]});
        }
    }
    
    // Return the dotProduct of two sparse vectors
    int dotProduct(SparseVector& vec) {
        int i=0, j=0, res=0;
        while(i<cc.size() && j < vec.cc.size()){
          if(cc[i].first == vec.cc[j].first)  res += cc[i].second * vec.cc[j].second, i++,j++;
          else if(cc[i].first < vec.cc[j].first)  i++; 
          else j++;
        }
        return res; 
    }
};

// Your SparseVector object will be instantiated and called as such:
// SparseVector v1(nums1);
// SparseVector v2(nums2);
// int ans = v1.dotProduct(v2);

